pkg/ottl/ottlfuncs/func_parse_csv.go

```diff
@@ -0,0 +1,172 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/parseutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+const (
+	parseCSVModeStrict       = "strict"
+	parseCSVModeLazyQuotes   = "lazyQuotes"
+	parseCSVModeIgnoreQuotes = "ignoreQuotes"
+)
+
+const (
+	parseCSVDefaultDelimiter = ','
+	parseCSVDefaultMode      = parseCSVModeStrict
+)
+
+type ParseCSVArguments[K any] struct {
+	Target          ottl.StringGetter[K]
+	Header          ottl.StringGetter[K]
+	Delimiter       ottl.Optional[string]
+	HeaderDelimiter ottl.Optional[string]
+	Mode            ottl.Optional[string]
+}
+
+func (p ParseCSVArguments[K]) validate() error {
+	if !p.Delimiter.IsEmpty() {
+		if len([]rune(p.Delimiter.Get())) != 1 {
+			return errors.New("delimiter must be a single character")
+		}
+	}
+
+	if !p.HeaderDelimiter.IsEmpty() {
+		if len([]rune(p.HeaderDelimiter.Get())) != 1 {
+			return errors.New("header_delimiter must be a single character")
+		}
+	}
+
+	return nil
+}
+
+func NewParseCSVFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ParseCSV", &ParseCSVArguments[K]{}, createParseCSVFunction[K])
+}
+
+func createParseCSVFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ParseCSVArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("ParseCSVFactory args must be of type *ParseCSVArguments[K]")
+	}
+
+	if err := args.validate(); err != nil {
+		return nil, fmt.Errorf("invalid arguments: %w", err)
+	}
+
+	delimiter := parseCSVDefaultDelimiter
+	if !args.Delimiter.IsEmpty() {
+		delimiter = []rune(args.Delimiter.Get())[0]
+	}
+
+	// headerDelimiter defaults to the chosen delimter,
+	// since in most cases headerDelimiter == delmiter.
+	headerDelimiter := delimiter
+	if !args.HeaderDelimiter.IsEmpty() {
+		headerDelimiter = []rune(args.HeaderDelimiter.Get())[0]
```

Comment:
Do we need to work with runes for the header? This code is executed at startup, so performance is not a concern

Reply:
Good point, we never use the header delimiter as runes so we don't need to cast it here like this.